File: src/frontend/src/lib/features/recipes/unitSpellings.ts
import type { BuiltInUnit } from './units';

/**
 * Folds a typed word to the form the table below is keyed by.
 *
 * German is written with umlauts and typed both ways: the same person writes
 * `Stück` at a keyboard and `Stueck` on a phone in a hurry. Folding here means
 * the table below holds one spelling of each word instead of two, and `Esslöffel`
 * cannot be the one that was forgotten.
 */
export const foldUnit = (word: string): string =>
  word
    .toLowerCase()
    .replace(/\.$/, '')
    .replaceAll('ä', 'ae')
    .replaceAll('ö', 'oe')
    .replaceAll('ü', 'ue')
    .replaceAll('ß', 'ss');

/**
 * What people actually type, mapped to the wire codes.
 *
 * The server reads the same words the same way (`Domain/Recipes/UnitSpellings`)
 * for units that arrive without passing through here — another app's import, a
 * unit typed into the editor. Here they are read before anything is sent, so a
 * pasted line is shown back already understood.
 *
 * Both languages, because a German recipe says `EL` and an English one says
 * `tbsp`, and the same person writes both depending on where the recipe came
 * from. Plurals are listed rather than stripped: German plurals are not a
 * suffix rule, and a parser that guessed would read `Zitronen` as a unit.
 */
const spellings: Record<string, BuiltInUnit> = {
  g: 'g',
  gr: 'g',
  gramm: 'g',
  gram: 'g',
  grams: 'g',
  gramme: 'g',
  kg: 'kg',
  kilo: 'kg',
  kilos: 'kg',
  kilogramm: 'kg',
  kilogram: 'kg',
  kilograms: 'kg',
  ml: 'ml',
  milliliter: 'ml',
  millilitre: 'ml',
  milliliters: 'ml',
  millilitres: 'ml',
  l: 'l',
  liter: 'l',
  litre: 'l',
  liters: 'l',
  litres: 'l',
  tsp: 'tsp',
  tsps: 'tsp',
  teaspoon: 'tsp',
  teaspoons: 'tsp',
  tl: 'tsp',
  teeloeffel: 'tsp',
  tbsp: 'tbsp',
  tbsps: 'tbsp',
  tbs: 'tbsp',
  tablespoon: 'tbsp',
  tablespoons: 'tbsp',
  el: 'tbsp',
  essloeffel: 'tbsp',
  stk: 'piece',
  stueck: 'piece',
  piece: 'piece',
  pieces: 'piece',
  zehe: 'clove',
  zehen: 'clove',
  clove: 'clove',
  cloves: 'clove',
  bund: 'bunch',
  bunches: 'bunch',
  bunch: 'bunch',
  scheibe: 'slice',
  scheiben: 'slice',
  slice: 'slice',
  slices: 'slice',
  dose: 'can',
  dosen: 'can',
  can: 'can',
  cans: 'can',
  packung: 'pack',
  packungen: 'pack',
  paeckchen: 'pack',
  pack: 'pack',
  packs: 'pack',
  packet: 'pack',
  packets: 'pack',
  prise: 'pinch',
  prisen: 'pinch',
  pinch: 'pinch',
  pinches: 'pinch'
};

/** The built-in unit a written word spells — `Milliliter`, `EL`, `Stk.` — if any. */
export const spelledUnit = (word: string): BuiltInUnit | undefined => spellings[foldUnit(word)];

File: src/backend/src/Domain/Recipes/UnitSpellings.cs
namespace Domain.Recipes;

/// <summary>
/// The words people write for the built-in units, in both languages.
/// </summary>
/// <remarks>
/// <para>
/// "Milliliter", "Gramm", "EL" and "Teelöffel" are not units a household made
/// up: they are <c>ml</c>, <c>g</c>, <c>tbsp</c> and <c>tsp</c> written out.
/// Kept as written, each would be a counting unit that never converts, never
/// sums with the built-in it means and shows its German name in an English
/// kitchen. Recipes arrive this way from every other app that stores units as
/// names, and cooks type them this way too.
/// </para>
/// <para>
/// Plurals are listed rather than stripped: German plurals are not a suffix
/// rule, and a rule that guessed would read <c>Zitronen</c> as a unit. The
/// frontend's line parser (<c>parseIngredientLine.ts</c>) reads the same
/// spellings out of a pasted line before anything reaches the server.
/// </para>
/// </remarks>
internal static class UnitSpellings
{
    private static readonly Dictionary<string, Unit> ByFold = new(StringComparer.Ordinal)
    {
        ["gr"] = Unit.Gram,
        ["gramm"] = Unit.Gram,
        ["gram"] = Unit.Gram,
        ["grams"] = Unit.Gram,
        ["gramme"] = Unit.Gram,
        ["kilo"] = Unit.Kilogram,
        ["kilos"] = Unit.Kilogram,
        ["kilogramm"] = Unit.Kilogram,
        ["kilogram"] = Unit.Kilogram,
        ["kilograms"] = Unit.Kilogram,
        ["milliliter"] = Unit.Millilitre,
        ["millilitre"] = Unit.Millilitre,
        ["milliliters"] = Unit.Millilitre,
        ["millilitres"] = Unit.Millilitre,
        ["liter"] = Unit.Litre,
        ["litre"] = Unit.Litre,
        ["liters"] = Unit.Litre,
        ["litres"] = Unit.Litre,
        ["tsps"] = Unit.Teaspoon,
        ["teaspoon"] = Unit.Teaspoon,
        ["teaspoons"] = Unit.Teaspoon,
        ["tl"] = Unit.Teaspoon,
        ["teeloeffel"] = Unit.Teaspoon,
        ["tbsps"] = Unit.Tablespoon,
        ["tbs"] = Unit.Tablespoon,
        ["tablespoon"] = Unit.Tablespoon,
        ["tablespoons"] = Unit.Tablespoon,
        ["el"] = Unit.Tablespoon,
        ["essloeffel"] = Unit.Tablespoon,
        ["stk"] = Unit.Piece,
        ["stueck"] = Unit.Piece,
        ["pieces"] = Unit.Piece,
        ["zehe"] = Unit.Clove,
        ["zehen"] = Unit.Clove,
        ["cloves"] = Unit.Clove,
        ["bund"] = Unit.Bunch,
        ["bunches"] = Unit.Bunch,
        ["scheibe"] = Unit.Slice,
        ["scheiben"] = Unit.Slice,
        ["slices"] = Unit.Slice,
        ["dose"] = Unit.Can,
        ["dosen"] = Unit.Can,
        ["cans"] = Unit.Can,
        ["packung"] = Unit.Pack,
        ["packungen"] = Unit.Pack,
        ["paeckchen"] = Unit.Pack,
        ["packs"] = Unit.Pack,
        ["packet"] = Unit.Pack,
        ["packets"] = Unit.Pack,
        ["prise"] = Unit.Pinch,
        ["prisen"] = Unit.Pinch,
        ["pinches"] = Unit.Pinch,
    };

    /// <summary>The built-in unit this word is a spelling of, or null.</summary>
    /// <param name="written">A unit as somebody wrote it, already trimmed.</param>
    internal static Unit? Resolve(string written) =>
        Unit.BuiltIn.FirstOrDefault(one => one.Code.Equals(written, StringComparison.OrdinalIgnoreCase))
            ?? ByFold.GetValueOrDefault(Fold(written));

    /// <summary>
    /// Lower case, no closing full stop, umlauts written out — so <c>Stück</c>,
    /// <c>Stueck</c> and <c>Stk.</c> all find their entry.
    /// </summary>
    private static string Fold(string written) =>
        written.TrimEnd('.').ToLowerInvariant()
            .Replace("ä", "ae", StringComparison.Ordinal)
            .Replace("ö", "oe", StringComparison.Ordinal)
            .Replace("ü", "ue", StringComparison.Ordinal)
            .Replace("ß", "ss", StringComparison.Ordinal);
}
